main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```
@@ -218,61 +221,58 @@ object Instances {
           }
       }
   }
 
   /**
     * Reassembles an instance
     */
-  private def checkInstance(inst: TypedAst.Instance, root: TypedAst.Root, changeSet: ChangeSet)(implicit flix: Flix): List[InstanceError] = {
-    checkSigMatch(inst, root) ::: checkOrphan(inst) ::: checkSuperInstances(inst, root)
+  private def checkInstance(inst: TypedAst.Instance, root: TypedAst.Root, changeSet: ChangeSet)(implicit sctx: SharedContext, flix: Flix): Unit = {
+    checkSigMatch(inst, root)
+    checkOrphan(inst)
+    checkSuperInstances(inst, root)
   }
 
   /**
     * Reassembles a set of instances of the same trait.
     */
-  private def checkInstancesOfTrait(insts0: List[TypedAst.Instance], root: TypedAst.Root, changeSet: ChangeSet)(implicit flix: Flix): List[InstanceError] = {
+  private def checkInstancesOfTrait(insts0: List[TypedAst.Instance], root: TypedAst.Root, changeSet: ChangeSet)(implicit sctx: SharedContext, flix: Flix): Unit = {
 
     // Instances can be uniquely identified by their heads,
     // due to the non-complexity rule and non-overlap rule.
     // This maps each instance head to its corresponding instance.
     var heads = Map.empty[TypeConstructor, TypedAst.Instance]
 
-    insts0.flatMap {
-      // check that the instance is on a valid type, suppressing other errors if not
-      case inst => checkSimple(inst) match {
-        // Case 1: No Errors. Run other checks.
-        case Nil =>
-          val res = checkOverlap(inst, heads) ::: checkInstance(inst, root, changeSet)
-          // add the head to the set
-          heads += (unsafeGetHead(inst) -> inst)
-          res
-        // Case 2: Errors. Skip other checks.
-        case errs => errs
-      }
+    insts0.foreach {
+      // check that the instance is on a valid type
+      case inst =>
+        checkSimple(inst)
+        checkOverlap(inst, heads)
+        checkInstance(inst, root, changeSet)
+        heads += (unsafeGetHead(inst) -> inst)
```

The semantic here is changed. Since checkSimple will no longer return a List, we can not rely on that to decide if we should do more checks.

If we want to preserve that semantic, we need to change the return type of checkSimple to something like a Boolean.

---

Ok, it seems that we have to preserve this semantic, or some tests will fail.

---

I think the reason we had the dependent checks is the `unsafeGetHead` function. It will probably crash if we have a complex instance (e.g. `instance Eq[a]`).

Maybe it's easiest if the `getHead` function just returns an option and we match on that?

---

So we can just use inst.tpe.typeConstructor to get the option. And it do eliminates most of the errors!

But there is still one test :
```
  test("Test.ComplexErrorSuppressesOtherErrors.01") {
    val input =
      """
        |trait C[a] {
        |  pub def f(x: a): Bool
        |}
        |
        |instance C[Box[String]] {
        |  pub def g(_x: Box[String]): Bool = false
        |}
        |
        |enum Box[a] {
        |    case Box(a)
        |}
        |""".stripMargin
    val result = compile(input, Options.TestWithLibNix)
    expectError[InstanceError.ComplexInstance](result)
    rejectError[InstanceError.ExtraneousDef](result)
  }
```
which is used to test if the error is suppressed. We may want to delete this test if we match over the option and do not suppress errors.